test/com/google/javascript/jscomp/CollapsePropertiesTest.java

```diff
@@ -202,11 +274,22 @@ public void testGlobalObjectNameInBooleanExpressionDepth3() {
   public void testGlobalFunctionNameInBooleanExpressionDepth1() {
     test("function a() {} a.c = 1; if (a) x(a.c);",
          "function a() {} var a$c = 1; if (a) x(a$c);");
+
+    test("function a() {} /** @nocollapse */ a.c = 1; if (a) x(a.c);",
+        "function a() {} a.c = 1; if (a) x(a.c);");
   }
 
   public void testGlobalFunctionNameInBooleanExpressionDepth2() {
     test("var a = {b: function(){}}; a.b.c = 1; if (a.b) x(a.b.c);",
          "var a$b = function(){}; var a$b$c = 1; if (a$b) x(a$b$c);");
+
+    test("var a = {/** @nocollapse */ b: function(){}}; a.b.c = 1; "
+        + "if (a.b) x(a.b.c);",
+        "var a = {b: function(){}}; var a$b$c = 1; if (a.b) x(a$b$c);");
```

Comment:
I've had cases where the property was valid to be renamed, but it was not exported. Specifically this came up using modules. Here's an example:

```JavaScript
// Code in module 1
var foo = {};
(function() {
  var a = 0; //some calculated value
  foo.bar = a;
})();
function showFooBar() {
  console.log(foo.bar);
}

// Code in module 2
foo.bar = 42;
showFooBar();
foo.bar = 0;
```

However, this is still more or less the same use case as exporting. I can come up with contrived examples where collapsing sub-properties is used, but not a single real world example.

The rationale behind implementing `@nocollapse` such that sub-properties were still collapsed was a fear of blocking further optimizations. However, perhaps that fear is unfounded.